File: 找出字符串中第一个匹配项的下标.cpp
class Solution {
public:
    int strStr(string main, string pattern) {
        int n = main.size(), m = pattern.size();
        string s = pattern + " " + main;
        vector<int> pi(n + m + 1);
        for (int i = 1; i <= n + m; ++i) {
            int len = pi[i - 1];
            while (len and s[i] != s[len]) {
                len = pi[len - 1];
            }
            if (s[i] == s[len]) {
                pi[i] = len + 1;
                if (pi[i] == m) {
                    return i - m * 2;
                }
            }
        }
        return -1; // 没找到
    }
};

File: 找出字符串中第一个匹配项的下标.py
class Solution:
    def strStr(self, main: str, pattern: str) -> int:
        n, m = len(main), len(pattern)
        s = pattern + " " + main
        pi = [0 for _ in range(n + m + 1)]
        for i in range(1, n + m + 1):
            l = pi[i - 1]
            while l and s[i] != s[l]:
                l = pi[l - 1]
            if s[i] == s[l]:
                pi[i] = l + 1
                if pi[i] == m:
                    return i - m * 2
        return -1
